// react-css-collapse/lib/components/Collapse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _util = require('../util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var initialStyle = {
  willChange: 'height',
  height: '0px',
  overflow: 'hidden'
};

var Collapse = function (_Component) {
  _inherits(Collapse, _Component);

  function Collapse() {
    _classCallCheck(this, Collapse);

    var _this = _possibleConstructorReturn(this, (Collapse.__proto__ || Object.getPrototypeOf(Collapse)).call(this));

    _this.onTransitionEnd = _this.onTransitionEnd.bind(_this);
    _this.setExpanded = _this.setExpanded.bind(_this);
    _this.setCollapsed = _this.setCollapsed.bind(_this);
    return _this;
  }

  _createClass(Collapse, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.content && this.props.isOpen) {
        this.setExpanded();
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var _this2 = this;

      if (!this.content) {
        return;
      }

      // expand
      if (!this.props.isOpen && nextProps.isOpen) {
        // have the element transition to the height of its inner content
        this.setContentStyleProperty('height', this.content.scrollHeight + 'px');
      }

      // collapse
      if (this.props.isOpen && !nextProps.isOpen) {
        // explicitly set the element's height to its current pixel height, so we
        // aren't transitioning out of 'auto'
        this.setContentStyleProperty('height', this.content.scrollHeight + 'px');
        _util2.default.requestAnimationFrame(function () {
          // "pausing" the JavaScript execution to let the rendering threads catch up
          // http://stackoverflow.com/questions/779379/why-is-settimeoutfn-0-sometimes-useful
          setTimeout(function () {
            _this2.setCollapsed();
          }, 0);
        });
      }
    }
  }, {
    key: 'onTransitionEnd',
    value: function onTransitionEnd(e) {
      var _props = this.props,
          onRest = _props.onRest,
          isOpen = _props.isOpen;


      if (isOpen) {
        this.setExpanded();
      }

      if (onRest && e.target === this.content && e.propertyName === 'height') {
        onRest();
      }
    }
  }, {
    key: 'setContentStyleProperty',
    value: function setContentStyleProperty(property, value) {
      this.content.style[property] = value;
    }
  }, {
    key: 'setCollapsed',
    value: function setCollapsed() {
      this.setContentStyleProperty('height', '0px');
      this.setContentStyleProperty('overflow', 'hidden');
    }
  }, {
    key: 'setExpanded',
    value: function setExpanded() {
      this.setContentStyleProperty('height', 'auto');
      this.setContentStyleProperty('overflow', 'visible');
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      return _react2.default.createElement(
        'div',
        {
          ref: function ref(el) {
            _this3.content = el;
          },
          style: initialStyle,
          className: this.props.className,
          onTransitionEnd: this.onTransitionEnd
        },
        this.props.children && this.props.children
      );
    }
  }]);

  return Collapse;
}(_react.Component);

Collapse.defaultProps = {
  isOpen: false,
  className: null,
  children: null,
  onRest: null
};

Collapse.propTypes = {
  children: _propTypes2.default.node,
  isOpen: _propTypes2.default.bool,
  className: _propTypes2.default.string,
  onRest: _propTypes2.default.func
};

exports.default = Collapse;
